File: Assets/Python/root/datastructs/datalist.py
from typing import Type, TypeVar, Generic, Any

from datastructs.command import Command, DirectCommand, CommandType
from datastructs.states import State
from networking.serializer import Serializer

T = TypeVar('T')
class Entry(Generic[T]):
    def __init__(self, message: str, data: T):
        self.message = message
        self.data = data

    @property
    def typename(self) -> str:
        return Serializer.type_to_name(type(self.data))

    def get_terminal_entry(target: str=None):
        command = Command(CommandType.EXIT) if target is None else DirectCommand(CommandType.EXIT, target)
        return Entry('terminate', command)

    def isTerminal(self, target: str=None):
        is_terminal_command = isinstance(self.data, Command) and self.data.type is CommandType.EXIT
        is_for_target = not isinstance(self.data, DirectCommand) or self.data.target == target
        return is_terminal_command and is_for_target
    
    def __str__(self) -> str:
        return f"({self.typename},{self.message},{self.data})"

class DataList:
    def __init__(self, headers:list[Entry]=None, content: list[Entry]=None):
        self.headers = headers or []
        self.content = content or []
    
    def add_headers(self, headers: list[Entry]):
        self.headers.extend(headers)

    def add_contents(self, content: list[Entry]):
        self.content.extend(content)

    def add_header(self, header: Entry):
        self.headers.append(header)

    def add_content(self, content: Entry):
        self.content.append(content)

    def get_state(self) -> State:
        entry = self.get_header_of_type(State)
        return entry.data if entry else State.unknown()
    
    def get_previous_state(self) -> State:
        entry: Entry[int] = self.get_header_by_message('previous state')
        return State(entry.data, True) if entry else State.unknown()
    
    def set_state(self, new_state: State):
        old = self.pop_header_of_type(State)
        self.pop_header_by_message('previous state')
        if old: self.add_header(Entry('previous state', old.data.id))
        self.add_header(Entry('current state', new_state))

    def set_previous_state(self, id: int):
        self.pop_header_by_message('previous state')
        self.add_header(Entry('previous state', id))
    
    def get_selected(self) -> list[str]:
        entry: Entry[str] = self.get_header_by_message('selected')
        if not entry: return []
        else: return entry.data.split(',')

    def set_selected(self, selected: list[str]):
        self.clear_selected()
        self.add_header(Entry('selected', ','.join(selected)))

    def clear_selected(self):
        self.pop_header_by_message('selected')
        
    def get_content_of_type(self, type: Type[T]) -> Entry[T]:
        return next((e for e in self.content if e.typename == Serializer.type_to_name(type)), None)
    
    def get_content_by_message(self, message: str):
        return next((e for e in self.content if e.message == message), None)
    
    def pop_content_of_type(self, type: Type[T]) -> Entry[T]:
        for e in self.content:
            if e.typename == Serializer.type_to_name(type):
                self.content.remove(e)
                return e
        return None
    
    def pop_content_by_message(self, *messages: str) -> Entry:
        for e in self.content:
            for m in messages:
                if e.message == m:
                    self.content.remove(e)
                    return e
        return None
    
    def pop_content_get_data_by_message(self, message: str) -> Any:
        entry = self.pop_content_by_message(message)
        return entry.data if entry else None
    
    def pop_content_to_tuple_by_messages(self, *messages: str) -> tuple:
        return tuple([self.pop_content_get_data_by_message(m) for m in messages])
    
    def get_header_of_type(self, type: Type[T]) -> Entry[T]:
        return next((e for e in self.headers if e.typename == Serializer.type_to_name(type)), None)
    
    def get_header_by_message(self, message: str):
        return next((e for e in self.headers if e.message == message), None)
    
    def pop_header_of_type(self, type: Type[T]) -> Entry[T]:
        for e in self.headers:
            if e.typename == Serializer.type_to_name(type):
                self.headers.remove(e)
                return e
        return None
    
    def pop_header_by_message(self, message: str) -> Entry:
        for e in self.headers:
            if e.message == message:
                self.headers.remove(e)
                return e
        return None
    
    
    def get_terminal(targets: list[str] = None):
        return DataList(headers=[Entry.get_terminal_entry(t) for t in (targets or [None])])
    
    def has_terminal_command_for(self, target: str = None):
        return any(e.isTerminal(target) for e in self.headers)
    
    def shallow_copy(self, new_state=None, clear_selected=False, new_contents=[]):
        copy = DataList(headers=self.headers[:], content=self.content[:])
        if new_state: copy.set_state(new_state)
        if clear_selected: copy.clear_selected()
        if new_contents: copy.add_contents(new_contents)
        return copy

    def __str__(self) -> str:
        return f"{list(map(str, self.headers))}{list(map(str, self.content))}"

File: Assets/Scripts/datastructs/DataList.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using Debug = UnityEngine.Debug;

public interface IEntry {
    public string GetTypeName();
    public string GetMessage();
    public object GetData();
    public bool IsTerminal(string target=null);
}

public class Entry<T> : IEntry {
    public string typename;
    public string message;
    public T data;

    public Entry(string message, T data) {
        this.message = message;
        this.data = data;
        typename = Serializer.TypeToName(data.GetType());
    }

    public static IEntry GetTerminal(string target=null) {
        var command = (target==null) ? new Command(CommandType.EXIT) : new DirectCommand(CommandType.EXIT, target); 
        return new Entry<Command>("terminate", command);
    }

    public object GetData() { return data; }
    public string GetMessage() { return message; }
    public string GetTypeName() { return typename; }

    public bool IsTerminal(string target=null) {
        bool isTerminal = (data is Command command) && (command.type == CommandType.EXIT);
        bool isForTarget = data is not DirectCommand dc || (dc.target == target);
        return isTerminal && isForTarget;
    }

    public override string ToString() {
        return $"({typename}, {message}, {data})";
    }
}

public class DataList {
    private List<IEntry> headers;
    private List<IEntry> content;

    public DataList(List<IEntry> headers=null, List<IEntry> content=null) {
        this.headers = headers ?? new();
        this.content = content ?? new();
    }
    public void AddHeaders(List<IEntry> headers) {
        this.headers.AddRange(headers);
    }
    public void AddContent(List<IEntry> content) {
        this.content.AddRange(content);
    }
    public void AddHeader<T>(Entry<T> header) {
        this.headers.Add(header);
    }
    public void AddContent<T>(Entry<T> content) {
        this.content.Add(content);
    }
    public void AddHeader(IEntry header) {
        this.headers.Add(header);
    }
    public void AddContent(IEntry content) {
        this.content.Add(content);
    }
    public List<IEntry> GetHeaders() {
        return headers;
    }
    public List<IEntry> GetContent() {
        return content;
    }
    public List<Entry<T>> GetContent<T>() {
        return content.Where(e => e.GetTypeName() == Serializer.TypeToName(typeof(T))).Select(e => (Entry<T>)e).ToList();
    }
    private Entry<T> Getter<T>(List<IEntry> list, string message=null) {
        Func<IEntry,bool> pred = (message==null) ? (e => e.GetTypeName() == Serializer.TypeToName(typeof(T))) : (e => e.GetMessage() == message); 
        return list.FirstOrDefault(pred) as Entry<T>;
    }
    private Entry<T> Popper<T>(List<IEntry> list, string message=null) {
        var entry = Getter<T>(list, message);
        if(entry != null) list.Remove(entry);
        return entry;
    }
    public Entry<T> ContentGetter<T>(string message=null) {return Getter<T>(content, message);}
    public Entry<T> HeaderGetter<T>(string message=null) {return Getter<T>(headers, message);}
    public Entry<T> ContentPopper<T>(string message=null) {return Popper<T>(content, message);}
    public Entry<T> HeaderPopper<T>(string message=null) {return Popper<T>(headers, message);}

    public State GetState() {
        return HeaderGetter<State>()?.data ?? State.UNKNOWN;
    }
    public State GetPreviousState() {
        return new((int?)(HeaderGetter<long>("previous state")?.data), true);
    }
    public void ChangeState(State newState) {
        var oldState = HeaderPopper<State>();
        HeaderPopper<long>("previous state");
        AddHeader<long>(new("previous state", oldState?.data.id ?? State.UNKNOWN.id));
        AddHeader<State>(new("current state", newState));
    }
    //AKA lying, not the best solution
    public void SetPreviousState(int id) {
        HeaderPopper<long>("previous state");
        AddHeader<long>(new("previous state", id));
    }

    public List<string> GetSelected() {
        return (HeaderGetter<string>("selected")?.data ?? "").Split(',', StringSplitOptions.RemoveEmptyEntries).ToList();
    }

    public void SetSelected(List<string> selected) {
        HeaderPopper<string>("selected");
        AddHeader<string>(new("selected", string.Join(",", selected)));
    }

    public static DataList GetTerminal(List<string> targets = null) {
        return new DataList(headers: (targets??new() {null}).Select(t => Entry<object>.GetTerminal(t)).ToList());
    }

    public bool HasTerminalCommand(string target = null) {
        return headers.Any(h => h.IsTerminal(target));
    }

    public DataList ShallowCopy(State newState=null, bool clearSelected=false) {
        DataList copy = new(new(headers), new(content));
        if(newState != null) copy.ChangeState(newState);
        if(clearSelected) copy.HeaderPopper<string>("selected");
        return copy;
    }

    public override string ToString() {
        return $"[{string.Join(",",headers)}][{string.Join(",",content)}]";
    }
}

